docs/Color-Picker/Color-Picker.md

```diff
@@ -1 +1,285 @@
+    * A static comparison of how the responses changed over time (the "results" screen)
+    * A dynamic comparison used while listening to a song (the live visualization)
+* The total audience distribution will at first contain only the concert submissions, but an admin should be able to update this to include additional data points.
+
+## API Contract
+
+### During Event
+#### Connect
+Send `REQUEST_AUTHENTICATE`
+```js
+{
+  token: String,    // access token from AWS Cognito
+}
+```
+
+#### Choose a color
+Send `COLOR_CHOSEN`
+```js
+{
+  color: String,
+  timestamp: String,
+}
+```
+
+#### Get chosen colors
+Send `REQUEST_MY_CHOICES`
+
+**Response**
+```js
+[
+  {
+    color: String,
+    timestamp: String,
+  },
+  ...
+]
+```
+
+#### Get aggregate choices
+Send `REQUEST_AGGREGATE_CHOICES`
+
+**Response**
+```js
+{
+  songs: [
+    {
+      id: String,
+      startTime: String,
+      endTime: Number,
+    },
+    ...
+  ],
+  choices: [
+    {
+      timestamp: Number,
+      colors: {
+        COLOR_RED: Number,
+        COLOR_BLUE: Number,
+        COLOR_GREEN: Number,
+        ...
+      },
+    },
+  ],
+},
+```
+
+#### Event: new choices aggregated
+Listen for `AGGREGATE_CHOICES_UPDATED`
+```js
+{
+  COLOR_RED: Number,
+  COLOR_BLUE: Number,
+  COLOR_GREEN: Number,
+  ...
+}
+```
+
+### After Event
+
+#### Connect
+Send `AUTHENTICATE`
+```js
+{
+  token: String,
+}
+```
+
+#### Start a new listen
+Send `REQUEST_CREATE_LISTEN`
+```js
+{
+  songId: String,
+  timestamp: String,
+}
+```
+
+**Response**
+```js
+{
+  id: String,
+  songId: String,
+  timestamp: String,
+}
+````
+
+#### Choose a color
+Send `COLOR_CHOSEN`
+```js
+{
+  listenId: String,
+  color: String,
+  timeOffset: Number,
+}
+```
+
+#### Get list of songs
+Send `REQUEST_GET_SONGS`
+
+**Response**
+```js
+[
+  {
+    id: String,
+    mediaUrl: String,
+    title: String,
+    artist: String,
+    length: Number,
+  }
+]
+```
+
+#### Get list of listens for a song
+Send `REQUEST_GET_LISTENS`
+```js
+{
+  songId: String,
+}
+```
+
+**Response**
+```js
+[
+  {
+    id: String,
+    songId: String,
+    timestamp: String,
+    choices: [
+      {
+        timeOffset: Number,
+        color: String,
+      }
+    ],
+  },
+  ...
+]
+```
+
+#### Get list of choices for a listen
+Send `REQUEST_GET_CHOICES`
+```js
+{
+  listendId: String,
+}
+```
+
+**Response**
+```js
+[
+  {
+    listenId: String,
+    color: String,
+    timeOffset: Number,
+  },
+  ...
+]
+```
+
+## Data Storage
+
+### During Event
+
+#### Song information
+```js
+{
+  songId: String,           // Partition key
+  listenId: 'SONG_INFO',    // Range key
+  title: String,            // Helpful for in-app displays
+  startTime: String,
+  endTime: String,
+}
+```
+
+#### User choices
+```js
+{
+  songId: String,           // Can be set to 'CONCERT_1' for the entire event
+  listenId: String,         // equal to userId
+  choices: {
+    [timestamp]: color,     // Key is the timestamp of the choice
```

Comment:
Here's the thinking behind making this an object with timestamp as the key instead of an array: if there's a disconnection or a network hiccup leads to a repeated request, we won't have duplicate entries.

Reply:
Makes sense to me!